scarlet/wavelet.py

```diff
@@ -28,91 +28,82 @@ def __init__(self, image, lvl = None, starlet = None, direct = False):
             if set to True, uses direct wavelet transform with the a trou algorithm.
             if set to False, the transform is performed by convolving the image by the wavelet transform of a dirac.
         """
-        # Shape for the starlet padding. It is also an fft fast shape.
         self.seed = None
-        if starlet is None:
-            self._starlet_shape = get_starlet_shape(image.shape, lvl = lvl)
-        else:
-            self._starlet_shape = starlet.shape
-        # Image (as a cube)
-        if len(image.shape) == 2:
-            self._image = image[np.newaxis, :, :]
-        else:
-            self._image = image
-        # Original shape of the image
-        self._image_shape = self._image.shape
-        # Can be initialised by a starlet transform to perform inverse transforms
-        if starlet is None:
-            if direct == True:
-                self._starlet = self.direct_transform()
+        # Transform method
+        self._direct = direct
+        if coefficients is None:
+            if image is None:
+                raise InputError('At least an image or a set of coefficients should be provided')
             else:
-                self._starlet = self.transform()
+                if len(image.shape) == 2:
+                    self._image = image[np.newaxis, :, :]
+                else:
+                    self._image = image
+                # Original shape of the image
+                self._image_shape = self._image.shape
+                # Padding shape for the starlet transform
+                if lvl is None:
+                    self._lvl = get_starlet_shape(image.shape)
+                else:
+                    self._lvl = lvl
+            self._coefficients = None
```

Comment:
I think this can be simplified. I'd always compute the coefficients in `__init__` and use  `@property coefficients` only to protect them from being overwritten.

The idea is: regardless of how the class is instantiated (with an image or coefficients), at the end of `__init__` it has the coefficients and forgets the image (if it ever had it).

Reply:
Then, I think what should happen is that the element (image or coefficient) the object is instantiated with should not be kept in memory: Once the inverse is computed, I don't really need the coefficients anymore..

Reply:
Fixed by updating image/coeffs whenever each other's setter is called.